TNOptimizer, can specify tags to optimize some tensors together

chris-n-self opened the pull request:
Added 'shared_tags' kwarg to TNOptimizer class, tensors with a `shared_tags` tag will be optimised together. Added a test to tests/test_tensor/test_optimizers.py to cover `shared_tags`

pep8speaks commented:
Hello @chris-n-self! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`quimb/tensor/optimize.py`](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py):

> [Line 128:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L128): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 161:41](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L161): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 162:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L162): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 163:21](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L163): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 163:22](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L163): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 163:64](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L163): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 163:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L163): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 169:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L169): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 79 characters)
> [Line 169:99](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L169): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 170:21](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L170): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 170:22](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L170): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 170:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L170): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 177:74](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L177): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 178:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L178): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 848:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/quimb/tensor/optimize.py#L848): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)

* In the file [`tests/test_tensor/test_optimizers.py`](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py):

> [Line 41:74](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L41): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 47:14](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L47): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '['
> [Line 47:17](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L47): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 47:23](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L47): [E228](https://duckduckgo.com/?q=pep8%20E228) missing whitespace around modulo operator
> [Line 47:44](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L47): [E202](https://duckduckgo.com/?q=pep8%20E202) whitespace before ']'
> [Line 52:32](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L52): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 52:44](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L52): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 61:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L61): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 64:80](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L64): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 84:13](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L84): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 84:19](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L84): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 271:36](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L271): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 272:6](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L272): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 272:12](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L272): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:22](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:28](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:31](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:36](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:42](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:45](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 274:49](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L274): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 286:22](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L286): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '['
> [Line 286:57](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L286): [E202](https://duckduckgo.com/?q=pep8%20E202) whitespace before ']'
> [Line 292:43](https://github.com/jcmgray/quimb/blob/b8d4bb587bf705b86a2f0ac570fc7293f163886b/tests/test_tensor/test_optimizers.py#L292): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

jcmgray commented on quimb/tensor/optimize.py:
```diff
@@ -147,6 +156,28 @@ def parse_network_to_backend(tn, tags, constant_tags, to_constant):
 
+        # if shared_tags are specified and this tensor contains one of them
+        if shared_tags and (t.tags & shared_tags):
+            if len(t.tags & shared_tags)>1:
```
The logic is getting a bit complex with all the various tags, I wonder if its worth doing this way (pseudocode-ish):
```python
if not opt_tags:
    # handle simple case of every tensor being a variable
    ...
    return

for t in tn.select(tags, 'any'):
    # handle normally tagged tensors
    ...

for tag in shared_tags:
    for t in tn.select(tag):
        # handle shared tags
        ...
```
but let me know if that seems ambitious for you to implement - I can always push some changes too.

jcmgray reviewed (commented):
Hey, thanks for the PR! Basically looks good! I've added a few comments, and it would be also be good if you could format the code for pep8, using a tool like e.g. `pylama`.

chris-n-self commented on tests/test_tensor/test_optimizers.py:
```diff
@@ -34,6 +35,55 @@
+
+    # tag circuit for shared_tags
+    tn_tagged = circ.psi.copy()
+    variable_finder = re.compile(r'ROUND_(\d+)')
+
+    for t in tn_tagged:
+        if ('RZZ' in t.tags):
+            for tag in t.tags:
+                match = variable_finder.match(tag)
+                if match is not None:
+                    i = int(match.groups(1)[0])
+                    t.add_tag('p'+f'{2*i}')
+        elif ('RX' in t.tags):
+            for tag in t.tags:
+                match = variable_finder.match(tag)
+                if match is not None:
+                    i = int(match.groups(1)[0])
+                    t.add_tag('p'+f'{2*i+1}')
```
Ah, this is perfect!

chris-n-self replied:
Is there a way to tag individual tensors when building a circuit? I tried putting a `tags=` in the call to `apply_gate` but that didn't work because the function inside `apply_gate` was already separately being passed a `tags` argument

chris-n-self commented on quimb/tensor/optimize.py:
```diff
@@ -728,6 +757,8 @@ class TNOptimizer:
     tags : str, or sequence of str, optional
         If supplied, only optimize tensors with any of these tags.
+    shared_tags : str, or sequence of str, optional
+        If supplied, optimize all tensors with these tags together
```
I have changed that using this language

jcmgray commented on quimb/tensor/optimize.py:
```diff
@@ -147,10 +153,70 @@ def parse_network_to_backend(tn, tags, constant_tags, to_constant):
+
+            # detect that this tensor is already variable tagged and skip
+            # if it is
+            if any([variable_finder.match(tag) for tag in t.tags]):
```
No need to change but thought I'd mention functions like `any` take a generator (i.e. you can remove the square brackets) - this also makes them more efficient since they lazily short circuit as soon as any True is found.

chris-n-self replied:
Ah, that's good to know, thanks!

jcmgray commented on tests/test_tensor/test_optimizers.py:
```diff
@@ -214,3 +252,33 @@ def test_multiloss(backend, executor):
+def test_shared_tags(tagged_qaoa_tn,backend):
+    n,depth,psi0 = tagged_qaoa_tn
+
+    H = qu.ham_heis(n,j=(0.,0.,-1.),b=(1.,0.,0.),cyclic=True,)
```
If possible, can you use pep8 style with spaces between arguments and operators etc.? Most editors can turn on such linting.

chris-n-self replied:
Sure, of course! I ran pylama on `optimise.py` but I forgot to run it on the test file. I haven't been able to get the pylint sublime plugin working yet

jcmgray commented on quimb/tensor/optimize.py:
```diff
@@ -147,10 +153,70 @@ def parse_network_to_backend(tn, tags, constant_tags, to_constant):
+            data = get_tensor_data(t)
+            variables.append(data)
+            t.add_tag(variable_tag.format(len(variables) - 1))
+        return tn_ag, variables
```
The new refactoring is clearer but I think one still needs to parse the `constant_tags` here. Sorry the logic might get a bit messy since there are essentially two styles of specifying tags now!

1. **Opt-out** - don't specify ``tags`` and maybe specify ``constant_tags`` (and maybe ``shared_tags`` - or should this trigger opt-in style?)
2. **Opt-in** - specify ``tags`` and maybe ``shared_tags`` - probably warn if ``constant_tags`` is supplied as redundant. Might be worth having those as two functions.

This might be factored as different functions to be even clearer.

If one was being most strict (and guessing least what user wants) one could even enforce that all ``shared_tags`` appear in ``tags``, I don't know what the best course is here - thoughts welcome.

Note for the opt-in case, one can likely use this :
```python
for t in tn.select_tensors(opt_tags, which='!any'):
    # iterate over tensors which *don't* have any of the given tags
    ... make_constant(t)
```
I probably need to document all the ways one can filter tensors somewhere...

chris-n-self replied:
Ahh, that is really cool that `!any` is a valid keyword for the `which`!

chris-n-self replied:
The opt-in/opt-out difference is a really helpful way to think of it!

> If one was being most strict (and guessing least what user wants) one could even enforce that all shared_tags appear in tags, I don't know what the best course is here - thoughts welcome.

That makes a lot of sense, and probably also helps to clarify to the user that in the opt_in case `tags` is the set of all tags that will be optimised over and then `shared_tags` is a subset of `tags` that will be optimised together. Otherwise the overlap between the two might be confusing. It feels potentially a bit strict to raise an error when a `shared_tag` doesn't appear in `tags` though. I can imagine that being quite a big source of frustration. Perhaps a warning and a comment that it has been automatically corrected? That way the intended use is made clear without introducing too much pain

jcmgray replied:
OK that sounds good. If ``tags`` or ``shared_tags`` are given assume opt-in mode, with ``shared_tags`` supposedly a subset of ``tags``. Else default to opt-out mode.

jcmgray reviewed (commented):
Appreciate the updates! - let me know if changes I'm proposing seem unclear or beyond what you  have time for - I can push some myself too.

jcmgray commented on tests/test_tensor/test_optimizers.py:
```diff
@@ -214,3 +252,33 @@ def test_multiloss(backend, executor):
+    assert len(tnopt.variables) == 2*depth 
+
+    psi_opt = tnopt.optimize_basinhopping(n=10, nhop=5)
+    assert sum(loss < -0.99 for loss in tnopt.losses) == 1
```
By the way would be good to test here a couple of things related to the shared tags functionality...
1. maybe that the TNOptimizer vector space is as expected smaller (i.e. 2 * 2 * depth?).
2. the data injected into ``psi_opt`` actually matches between shared tags tensors

rezah replied:
Few comments here: I tested it for a periodic MPS, as it worked. I tried different scenarios also (same tensors in the bulk) still works well. 

As a QUIMB user, I think share-tags and tags should be totally different (no common elements), that at the least for me makes more sense as I see what I am targeting. 

Maybe const_tags are redundant here makes things more complicated.

rezah replied:
The problem I am facing is that results are not consistant. See the following example: periodic MPS with one variable and multiple variable for a periodic Hamiltonian. It seems in periodic MPS with one common tensor, it can not auto-diff correctly as energy is much higher than the same MPS with multiple tensors.   



`
%matplotlib inline
from quimb import *
from quimb.tensor import *
import quimb as qu
import quimb.tensor as qtn
import numpy as np
from orderedset import OrderedSet
L = 6
psi = MPS_rand_state(L, 12, phys_dim=2, normalize=True, cyclic=True, seed=10)
psi.show()
ham = MPO_ham_heis(L, j=(1.0,1.0,1.0), bz=0.0, S=0.5, cyclic=True)

E_exact = qu.heisenberg_energy(L)
print ("energy_exact", L, E_exact)

for i in range(L):
 psi[i].add_tag("I")


print (psi)


def norm_mps(psi):
    norm_val=(psi.H & psi).contract()
    psi=psi*(norm_val**(-0.5))
    return psi

def energy_f(psi, ham):
    p_H=psi.H 
    psi.align_(ham, p_H)
    return ( p_H & ham & psi).contract()


def loss_f(psi, p_DMRG):
    return  1-abs((psi & p_DMRG.H).contract(all, optimize='auto-hq'))

tnopt = qtn.TNOptimizer(
    psi,                         
    loss_fn=energy_f,                       
    norm_fn=norm_mps,                         
    loss_constants={ "ham" : ham },  
    tags={  },                 
    shared_tags={"I"},
    constant_tags={ },                   
    optimizer='adam',             
    autodiff_backend='torch',         
)

psi=tnopt.optimize(500, tol=1e-12, ftol=1e-12, gtol=1e-12)


psi = MPS_rand_state(L, 6, phys_dim=2, normalize=False, cyclic=True)
for i in range(L):
 psi[i].add_tag("I")


tnopt = qtn.TNOptimizer(
    psi,                         
    loss_fn=energy_f,                       
    norm_fn=norm_mps,                         
    loss_constants={ "ham" : ham },  
    tags={ "I" },                 
    shared_tags={},
    constant_tags={ },                   
    optimizer='L-BFGS-B',             
    autodiff_backend='torch',         
)

psi=tnopt.optimize(220, tol=1e-12, ftol=1e-12, gtol=1e-12)

`

jcmgray replied:
I tried this with even and odd tensor with shared data and the energy was good - I think something to do with having a large enough unit cell rather than the autodiff implementation no?

rezah replied:
Ah, smart. The Heisenberg model has an anti-Ferro magnet structure requiring a two-unit cell ...AB...probably that is the reason.

Seems it work well :)

chris-n-self commented:
I have refactored things so the `parse_network_to_backend` function either internally calls an opt-in parser, or an opt-out parser. Since it's more complicated now I added a bit of an explanatory docstring to `parse_network_to_backend`

rezah commented:
@chris-n-self I think in both cases, (especially in opt_in) `constant_tags` should not be ignored. They can still play an important role in making things simple for user.

jcmgray commented:
Thanks for the changes @chris-n-self, will hopefully look at today.

@rezah Do you mean you'd like to specify tags marking a sub-set of tensors to optimize, then a sub-sub-set of those tensors *not* to optimize? I think the logic here was you can just untag the constant tensors instead.

rezah commented:
@jcmgray Yes. I agree maybe in the way you are pointing out (untag constants tensors) it is more clear and less confusing.

chris-n-self commented:
I agree with Johnnie, I think if there are too many overlapping subsets then what rules take precedence could get confusing, i.e. if something has both an optimisation tag and a constant_tag then you would have to remember that constant_tag overrules the optimisation tag, rather than the other way around

jcmgray commented:
All looks good to me, thanks for great PR @chris-n-self! I'm going to start an issue collating various ways the docs might be improved, but a page on the `TNOptimizer` is definitely overdue if either of you have suggestions on what might be useful to include.
